src/provider/states/pod/starting.rs

```diff
@@ -1,122 +1,145 @@
-                        Ok(true) => debug!(
-                            "Service [{}] still running after [{}] seconds",
-                            &unit.get_name(),
-                            start_time.elapsed().as_secs()
-                        ),
-                        Ok(false) => {
-                            return Transition::Complete(Err(anyhow!(format!(
-                                "Unit [{}] stopped unexpectedly during startup after [{}] seconds.",
-                                &unit.get_name(),
-                                start_time.elapsed().as_secs()
-                            ))))
-                        }
-                        Err(dbus_error) => return Transition::Complete(Err(dbus_error)),
-                    }
-                }
+        let pod = pod.latest();
+
+        match start_service_units(shared, pod_state, &pod).await {
+            Ok(()) => Transition::next(self, Running::default()),
+            Err(error) => {
+                error!("{}", error);
+                Transition::Complete(Err(error))
             }
-        } else {
-            warn!(
-                "No unit definitions found, not starting anything for pod [{}]!",
-                pod_state.service_name
-            );
         }
-        Transition::next(
-            self,
-            Running {
-                ..Default::default()
-            },
+    }
+
+    async fn status(&self, _pod_state: &mut PodState, _pod: &Pod) -> Result<PodStatus> {
+        Ok(make_status(Phase::Pending, "Starting"))
+    }
+}
+
+/// Starts the service units for the containers of the given pod.
+///
+/// The units are started and enabled if they are not already running.
+/// The startup is considered successful if the unit is still running
+/// after 10 seconds.
+async fn start_service_units(
+    shared: SharedState<ProviderState>,
+    pod_state: &PodState,
+    pod: &Pod,
+) -> Result<()> {
+    let pod_key = &PodKey::from(pod);
+
+    let (systemd_manager, pod_handle) = {
+        let provider_state = shared.read().await;
+        let handles = provider_state.handles.read().await;
+        (
+            provider_state.systemd_manager.clone(),
+            handles.get(&pod_key).map(PodHandle::to_owned),
         )
+    };
+
+    for (container_key, container_handle) in pod_handle.unwrap_or_default() {
+        let service_unit = &container_handle.service_unit;
+
+        if systemd_manager.is_running(service_unit)? {
+            debug!(
+                "Unit [{}] for service [{}] is already running. Skip startup.",
+                service_unit, &pod_state.service_name
+            );
+        } else {
+            info!("Starting systemd unit [{}]", service_unit);
+            systemd_manager.start(service_unit)?;
+
+            info!("Enabling systemd unit [{}]", service_unit);
+            systemd_manager.enable(service_unit)?;
+
+            // TODO: does this need to be configurable, or ar we happy with a hard coded value
```

Comment:
Yes, this can be improved like you mentioned. I just rearranged this code but did not change it because the pull request is large enough and I did not want to touch things not related to logging (the rearrangement was related).

Reply:
Totally fine, I'll copy my comment to an issue so we don't forget about it.